# Exploring the Pokemon Dataset

#### **SETUP**

In [27]:
#importing relevant modules

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### **CHECKING INPUT DATA**

In [66]:
#Reading the csv file and printing first few rows of data
data = pd.read_csv('pokemon.csv')

#Let's work with a copy of the data.
df = data.copy()
df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Okay do there's inconcsistencies with the 'Name' and '#' columns. The name after Venusaur looks sus and also has the wrong value in the '#' column.

Hmm. We'll check out the 'Name' and the '#' columns and see if this is something that's happening throughout. We'll get to formatting them once we do some preliminary checks.

#### **CHECKING FOR DUPLICATES**

In [67]:
df.loc[df.duplicated()]

Empty DataFrame
Columns: [#, Name, Type 1, Type 2, Total, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed, Generation, Legendary]
Index: []

In [68]:
df.loc[df.duplicated('Name')]

Empty DataFrame
Columns: [#, Name, Type 1, Type 2, Total, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed, Generation, Legendary]
Index: []

Great! Looks like there aren't any duplicate rows in the data.

#### **CHECKING DATA TYPES**

In [69]:
#Info on all columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


Nice! The data types are all okay.

#### **LOOKING FOR NULL VALUES**

In [ ]:
#Check to see if there are any null values
df .isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

Okay. So the 'Type 2' column has some null values. Let's see what to do with them in a while. In order ot have a cleaner dataframe to work with, let's ust go ahead and fix the '#' and 'Name' columns first.

#### **FIXING THE '#' COLUMN**

It looks like this column is just meant to serve as an index and nothing else. But, it has wrong values for many rows where the number repeats. Since it doesn't serve any other purpose or has any bearing on our data, it is safe to just drop it.

In [ ]:
#Dropping '#' column
df = df.drop('#', axis = 1)

#Check to see if the columns is dropped
#df

Name    Type 1    Type 2  Total   HP  Attack  \
0                    Bulbasaur     Grass    Poison    318   45      49   
1                      Ivysaur     Grass    Poison    405   60      62   
2                     Venusaur     Grass    Poison    525   80      82   
3        VenusaurMega Venusaur     Grass    Poison    625   80     100   
4                   Charmander      Fire       NaN    309   39      52   
5                   Charmeleon      Fire       NaN    405   58      64   
6                    Charizard      Fire    Flying    534   78      84   
7    CharizardMega Charizard X      Fire    Dragon    634   78     130   
8    CharizardMega Charizard Y      Fire    Flying    634   78     104   
9                     Squirtle     Water       NaN    314   44      48   
10                   Wartortle     Water       NaN    405   59      63   
11                   Blastoise     Water       NaN    530   79      83   
12     BlastoiseMega Blastoise     Water       NaN    630   79     103   
13                    Caterpie       Bug       NaN    195   45      30   
14                     Metapod       Bug       NaN    205   50      20   
15                  Butterfree       Bug    Flying    395   60      45   
16                      Weedle       Bug    Poison    195   40      35   
17                      Kakuna       Bug    Poison    205   45      25   
18                    Beedrill       Bug    Poison    395   65      90   
19       BeedrillMega Beedrill       Bug    Poison    495   65     150   
20                      Pidgey    Normal    Flying    251   40      45   
21                   Pidgeotto    Normal    Flying    349   63      60   
22                     Pidgeot    Normal    Flying    479   83      80   
23         PidgeotMega Pidgeot    Normal    Flying    579   83      80   
24                     Rattata    Normal       NaN    253   30      56   
25                    Raticate    Normal       NaN    413   55      81   
26                     Spearow    Normal    Flying    262   40      60   
27                      Fearow    Normal    Flying    442   65      90   
28                       Ekans    Poison       NaN    288   35      60   
29                       Arbok    Poison       NaN    438   60      85   
30                     Pikachu  Electric       NaN    320   35      55   
31                      Raichu  Electric       NaN    485   60      90   
32                   Sandshrew    Ground       NaN    300   50      75   
33                   Sandslash    Ground       NaN    450   75     100   
34                    Nidoran♀    Poison       NaN    275   55      47   
35                    Nidorina    Poison       NaN    365   70      62   
36                   Nidoqueen    Poison    Ground    505   90      92   
37                    Nidoran♂    Poison       NaN    273   46      57   
38                    Nidorino    Poison       NaN    365   61      72   
39                    Nidoking    Poison    Ground    505   81     102   
40                    Clefairy     Fairy       NaN    323   70      45   
41                    Clefable     Fairy       NaN    483   95      70   
42                      Vulpix      Fire       NaN    299   38      41   
43                   Ninetales      Fire       NaN    505   73      76   
44                  Jigglypuff    Normal     Fairy    270  115      45   
45                  Wigglytuff    Normal     Fairy    435  140      70   
46                       Zubat    Poison    Flying    245   40      45   
47                      Golbat    Poison    Flying    455   75      80   
48                      Oddish     Grass    Poison    320   45      50   
49                       Gloom     Grass    Poison    395   60      65   
50                   Vileplume     Grass    Poison    490   75      80   
51                       Paras       Bug     Grass    285   35      70   
52                    Parasect       Bug     Grass    405   60      95   
53                     Venonat

Nice. Let's now fix the 'Name' column.

#### **WORKING WITH 'NAME' COLUMN**

In [72]:
#Display all row instead of just the first and last few
#pd.set_option('display.max_rows', None)

#Checking out the 'Name and '#' column values
#print(df['Name'])

Okay. Clearly, the names have issues in them. And it seems to be specifically ones with two words in them.

Let's just look at all names which have two words in them. And then, we'll see how to procees.

In [73]:
#Seeing all the problem names in them (ones with two words)
#df.loc[df['Name'].str.contains( pat = r'(\w+) (\w+)', regex = True)]

Awesome. Right off the bat, we see that all Pokemon with 'Mega' in their name have a redundant name at the beginning. Let's remove that.

In [74]:
#Removing redundant name in names with 'Mega'
df['Name'] = df['Name'].replace(to_replace = r'(\w+)(Mega)',value = r'\2', regex = True)

#Check to see if all of them are fixed
#df.loc[df['Name'].str.contains('Mega')]

In [75]:
#Looking up the column again to see what to fix next
#df['Name']

Great! The 'Mega' Pokemon are all good. 

On scrolling through the 'Name' column, we also see that two Pokemon with 'Primal' in their name have the same issue. Let's fix that up too.

In [76]:
#Removing redundant name in names with 'Primal'
df['Name'] = df['Name'].replace(to_replace = r'(\w+)(Primal)', value = r'\2', regex = True)

#Check to see if all of them are fixed
#df.loc[df['Name'].str.contains('Primal')]

Great! The 'Primal' Pokemon are fixed too.

Let's see what to do next.

In [77]:
#df['Name']

Okay. So there's one more Pokemon which seems to have it's name repeating at the beginning and the end - 'Rotom'. And it also seems like there are many Pokemon who's names have their first word as a concatenation of two words, alongwith one that has a concatenation of a word and some numbers (Zygarde50% Forme). 

Let's quickly fix the Rotoms up before moving on to split the names where there is a concatenation of words.

In [78]:
#Replace 
df['Name'] = df['Name'].replace(to_replace = r'(Rotom)([A-Z])', value = r'\2', regex = True)

#Check to see if this fixed all Rotom Pokemon names
#df.loc[df['Name'].str.contains('Rotom')]

Amazing!! Finally, let's fix those Pokemon who's first name is a concatenation of two words by adding a space in between them.

In [79]:
#Add space in between the words
df['Name'] = df['Name'].replace(to_replace = r'([a-z])([A-Z]|\d)', value = r'\1 \2', regex = True)

#Check the dataset
#df.loc[df['Name'].str.contains(r'([a-z]) ([A-Z]|\d)', regex = True)]

Oh it looks like there's another Pokemon with a duplicate name : 'Kyurem Black Kyurem' and 'Kyurem White Kyurem'. That needs fixing.

Also, after a quick internet search, apparently 'Hoopa Hoopa Confined' and 'Hoopa Hoopa Unbound' are just 'Hoopa Confined' and 'Hoopa Unbound'. 

Let's fix all this.

In [80]:
#Fixing Kyurem
df['Name'] = df['Name'].replace(to_replace = r'(Kyurem) ([A-Z])', value = r'\2', regex = True)

#Fixing Hoopa
df['Name'] = df['Name'].replace(to_replace = r'(Hoopa) (Hoopa)', value = r'\2', regex = True)

In [81]:
#Final look (hopefully)
#df

AMAZING! All Pokemon names have been fixed. We can move on now.

#### **FIXING NULL VALUES**

Let's go ahead and fix the null values in the 'Type 2' Column.

Oh okay. So after a quick internet search, apparently, Pokemon with two types are called 'Dual-type' Pokemon and the ones with a single type are called 'Pure-type' Pokemon. So, the ones with NaN in their 'Type 2' column are Pure-type pokemon. 

Okay so keeping that in mind, I'm actually going to leave the column as is and not think of filling up the null values. In case ther are issues later, I will come back to this and fill up the values accordingly.

You know what, let's also add a column that tells us what type (Dual Type/Pure Type) the Pokemon is.

In [ ]:
#Initializing the 'Type' Column with 'Dual Type'
df.insert(1, 'Type', 'Dual Type')

#If 'Type 2' column is null, set 'Type' column to be 'Pure Type'
df.loc[df['Type 2'].isnull(), 'Type'] = 'Pure Type'

Name       Type    Type 1    Type 2  Total   HP  \
0                    Bulbasaur  Dual Type     Grass    Poison    318   45   
1                      Ivysaur  Dual Type     Grass    Poison    405   60   
2                     Venusaur  Dual Type     Grass    Poison    525   80   
3                Mega Venusaur  Dual Type     Grass    Poison    625   80   
4                   Charmander  Pure Type      Fire       NaN    309   39   
5                   Charmeleon  Pure Type      Fire       NaN    405   58   
6                    Charizard  Dual Type      Fire    Flying    534   78   
7             Mega Charizard X  Dual Type      Fire    Dragon    634   78   
8             Mega Charizard Y  Dual Type      Fire    Flying    634   78   
9                     Squirtle  Pure Type     Water       NaN    314   44   
10                   Wartortle  Pure Type     Water       NaN    405   59   
11                   Blastoise  Pure Type     Water       NaN    530   79   
12              Mega Blastoise  Pure Type     Water       NaN    630   79   
13                    Caterpie  Pure Type       Bug       NaN    195   45   
14                     Metapod  Pure Type       Bug       NaN    205   50   
15                  Butterfree  Dual Type       Bug    Flying    395   60   
16                      Weedle  Dual Type       Bug    Poison    195   40   
17                      Kakuna  Dual Type       Bug    Poison    205   45   
18                    Beedrill  Dual Type       Bug    Poison    395   65   
19               Mega Beedrill  Dual Type       Bug    Poison    495   65   
20                      Pidgey  Dual Type    Normal    Flying    251   40   
21                   Pidgeotto  Dual Type    Normal    Flying    349   63   
22                     Pidgeot  Dual Type    Normal    Flying    479   83   
23                Mega Pidgeot  Dual Type    Normal    Flying    579   83   
24                     Rattata  Pure Type    Normal       NaN    253   30   
25                    Raticate  Pure Type    Normal       NaN    413   55   
26                     Spearow  Dual Type    Normal    Flying    262   40   
27                      Fearow  Dual Type    Normal    Flying    442   65   
28                       Ekans  Pure Type    Poison       NaN    288   35   
29                       Arbok  Pure Type    Poison       NaN    438   60   
30                     Pikachu  Pure Type  Electric       NaN    320   35   
31                      Raichu  Pure Type  Electric       NaN    485   60   
32                   Sandshrew  Pure Type    Ground       NaN    300   50   
33                   Sandslash  Pure Type    Ground       NaN    450   75   
34                    Nidoran♀  Pure Type    Poison       NaN    275   55   
35                    Nidorina  Pure Type    Poison       NaN    365   70   
36                   Nidoqueen  Dual Type    Poison    Ground    505   90   
37                    Nidoran♂  Pure Type    Poison       NaN    273   46   
38                    Nidorino  Pure Type    Poison       NaN    365   61   
39                    Nidoking  Dual Type    Poison    Ground    505   81   
40                    Clefairy  Pure Type     Fairy       NaN    323   70   
41                    Clefable  Pure Type     Fairy       NaN    483   95   
42                      Vulpix  Pure Type      Fire       NaN    299   38   
43                   Ninetales  Pure Type      Fire       NaN    505   73   
44                  Jigglypuff  Dual Type    Normal     Fairy    270  115   
45                  Wigglytuff  Dual Type    Normal     Fairy    435  140   
46                       Zubat  Dual Type    Poison    Flying    245   40   
47                      Golbat  Dual Type    Poison    Flying    455   75   
48                      Oddish  Dual Type     Grass    Poison    320   45   
49                       Gloom  Dual Type     Grass    Poison    395   60   
50                   Vileplume  Dual Type     Grass    Poison    490   75   
51                    

In [ ]:
#Let's check the null values again
df.isnull().sum()

Name            0
Type            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

In [87]:
#Let's check all of the Pure type pokemon
df.loc[df['Type'] == 'Pure Type'].shape

(386, 13)

Awesome! We had 386 null values initially, and so, can infer that all of them have been ascribed to 'Pure Type' Pokemon.

#### **DATA STATS**

In [61]:
#Let's see the stats of the data
df.describe()

Total          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.00000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   435.10250   69.258750   79.001250   73.842500   72.820000   71.902500   
std    119.96304   25.534669   32.457366   31.183501   32.722294   27.828916   
min    180.00000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    330.00000   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    450.00000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    515.00000   80.000000  100.000000   90.000000   95.000000   90.000000   
max    780.00000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000000     6.00000

Okay. So, we have a total of 800 Pokemon spanning 6 generations.

**DIFFERENT TYPES OF POKEMON**

In [ ]:
#All Type 1 values
print(f'{df['Type 1'].unique()} : {len(df['Type 1'].unique())} types.')

#All Type 2 values
print(f'{df['Type 2'].unique()} : {len(df['Type 2'].unique())} types.')

['Grass' 'Fire' 'Water' 'Bug' 'Normal' 'Poison' 'Electric' 'Ground'
 'Fairy' 'Fighting' 'Psychic' 'Rock' 'Ghost' 'Ice' 'Dragon' 'Dark' 'Steel'
 'Flying'] : 18 types.
['Poison' nan 'Flying' 'Dragon' 'Ground' 'Fairy' 'Grass' 'Fighting'
 'Psychic' 'Steel' 'Ice' 'Rock' 'Dark' 'Water' 'Electric' 'Fire' 'Ghost'
 'Bug' 'Normal'] : 19 types.


Looks like both Type 1 and Type 2 have similar types, except for the null values. Let's check that quickly.

If all the elements in 'Type 2' exist in 'Type 1' except for the null value, then we will be right. So, we'll look for all the values that are in Type 2 but not in Type 1. If we are right, it should return a list containing only a null.

In [ ]:
#Check if all elements in Type 2 are in Type 1
print(df.loc[~df['Type 2'].isin(df['Type 1']), 'Type 2'].unique())

#Number of elements
print(len(df.loc[~df['Type 2'].isin(df['Type 1']), 'Type 2']))

[nan]
386


Cool! So we see that the 'Type 2' contains the same set of values as 'Type 1', but in a different order of course. THe only exceptions are the null values which correspond to the Pure-Type Pokemon, as we saw before.

From here on, let's work with the Dual type pokemon first, and then, the Pure type pokemon.

#### **DUAL TYPE POKEMON**

In [146]:
#Total number of each type of pokemon as per their Type 1
#df.loc[(df['Type'] == 'Dual Type'), ['Type 1']].value_counts()

In [147]:
#Total number of pokemon as categorised by both type 1 and type 2
df.loc[(df['Type'] == 'Dual Type'), ['Type 1', 'Type 2']].value_counts()

Type 1    Type 2  
Normal    Flying      24
Grass     Poison      15
Bug       Flying      14
          Poison      12
Ghost     Grass       10
Water     Ground      10
Steel     Psychic      7
Fire      Fighting     7
Water     Flying       7
Bug       Steel        7
Psychic   Flying       6
          Fairy        6
Water     Dark         6
Rock      Ground       6
Dragon    Flying       6
Fire      Flying       6
Rock      Water        6
Bug       Grass        6
Dragon    Ground       5
Electric  Flying       5
Normal    Fairy        5
Dark      Flying       5
Water     Psychic      5
Grass     Flying       5
Ghost     Poison       4
Steel     Ghost        4
Water     Rock         4
Rock      Flying       4
Dragon    Psychic      4
Ground    Flying       4
Grass     Dark         3
          Ice          3
Fire      Ground       3
Poison    Flying       3
Rock      Fairy        3
Ghost     Fire         3
Psychic   Fighting     3
Fighting  Psychic      3
Ground    Rock         3
      

In [148]:
#Verifying if the above output is correct
#df.loc[(df['Type'] == 'Dual Type') & (df['Type 1'] == 'Bug'), ['Type 1', 'Type 2']].value_counts().sum()

# **CONTINUE HERE**

In [ ]:
#Graph of the types of Pokemon in Dual Type
#Grpah of attack, defense, speed, ap. attack. sp. defense, HP, total wrt generation, legendary 

In [ ]:
#Do same for Pure type pokemon